libpod/container_api.go

```diff
@@ -834,3 +840,181 @@ func (c *Container) Refresh(ctx context.Context) error {
+		networkJson, err := ioutil.ReadAll(networkStatusFile)
+		if err != nil {
+			return err
+		}
+		json.Unmarshal(networkJson, &networkStatus)
+		// Take the first IP address
+		var IP net.IP
+		if len(networkStatus) > 0 {
+			if len(networkStatus[0].IPs) > 0 {
+				IP = networkStatus[0].IPs[0].Address.IP
+			}
+		}
+		if IP != nil {
+			env := fmt.Sprintf("IP=%s", IP)
+			// Tell CNI which IP address we want.
+			os.Setenv("CNI_ARGS", env)
+			logrus.Debugf("Restoring container with %s", env)
+		}
+	}
+
+	if err := c.prepare(); err != nil {
+		return err
+	}
+	defer func() {
+		if err != nil {
+			if err2 := c.cleanup(); err2 != nil {
+				logrus.Errorf("error cleaning up container %s: %v", c.ID(), err2)
+			}
+		}
+	}()
+
+	// CNI_ARGS was used to request a certain IP address. Unconditionally remove it.
+	os.Unsetenv("CNI_ARGS")
+
+	// Read config
+	jsonPath := filepath.Join(c.bundlePath(), "config.json")
+	logrus.Debugf("generate.NewFromFile at %v", jsonPath)
+	g, err := generate.NewFromFile(jsonPath)
+	if err != nil {
+		logrus.Debugf("generate.NewFromFile failed with %v", err)
+		return err
+	}
+
+	// We want to have the same network namespace as before.
+	if c.config.CreateNetNS {
+		g.AddOrReplaceLinuxNamespace(spec.NetworkNamespace, c.state.NetNS.Path())
+	}
+
+	// Save the OCI spec to disk
+	if err := c.saveSpec(g.Spec()); err != nil {
+		return err
+	}
+
+	// Copy /etc/resolv.conf to the container's rundir
+	_, err = c.generateResolvConf()
+	if err != nil {
+		return err
+	}
+
+	// Copy /etc/hosts to the container's rundir
+	_, err = c.generateHosts()
+	if err != nil {
+		return errors.Wrapf(err, "unable to copy /etc/hosts to container space")
+	}
+
+	// This needs to be deleted so that the restore works correctly.
+	// This also happens in Chechkpoint()
+	// Not sure where is the right place for this.
+	// If the container is stopped via runc, it needs to be here for a successful restart.
+	os.Remove(filepath.Join(c.bundlePath(), "attach"))
+	os.Remove(filepath.Join(c.bundlePath(), "ctl"))
+
+	if err := c.runtime.ociRuntime.createOCIContainer(c, c.config.CgroupParent, true); err != nil {
+		return err
+	}
+
+	logrus.Debugf("Restored container %s", c.ID())
+
+	c.state.State = ContainerStateRunning
+
+	if keep {
+		return c.save()
+	}
+
+	// Delete all checkpoint related files. At this point, in theory, all files
+	// should exist. Still ignoring errors for now as the container should be
+	// restored and running. Not erroring out just because some cleanup operation
+	// failed. Starting with the checkpoint directory
+	os.RemoveAll(filepath.Join(c.bundlePath(), "checkpoint"))
+	// Log files
+	os.Remove(filepath.Join(c.bundlePath(), "restore.log"))
+	os.Remove(filepath.Join(c.bundlePath(), "dump.log"))
+	// Statistic files
+	os.Remove(filepath.Join(c.bundlePath(), "stats-dump"))
+	os.Remove(filepath.Join(c.bundlePath(), "stats-restore"))
+	// Network status file
+	os.Remove(filepath.Join(c.bundlePath(), "network.status"))
```

Comment:
I concur with the no error out, but am just wondering if throwing a debug message on error might be helpful if things ever go wrong.  I think it's probably fine as is, but thought I'd raise the question for consideration.

Reply:
Sounds like a good idea.